### PR TITLE
Issues with `foundry-kompile` colliding with K internals

### PR DESCRIPTION
**Pull request by iFrostizz**
Closes https://github.com/runtimeverification/evm-semantics/issues/1368, closes https://github.com/runtimeverification/evm-semantics/issues/1703

Now supports:

- [ ] Single name contracts by escaping by adding the ~`KEVM`~ `S2K` prefix
- [ ] Contracts name with an underscore by escaping (adds the prefix + replace the underscore by its ASCII representation, can also unescape, see tests)
- [ ] Method names such as `claim()` won't clash anymore with a K claim, we ~add `method-` in front of it~ **escape everything**

For visualization, consider a project structure such as:

```sh
.
├── cache
│   └── solidity-files-cache.json
├── foundry.toml
├── out
│   ├── Counter.sol
│   │   ├── Counter.json
│   │   └── COuntER.json
│   ├── Counter.t.sol
│   │   ├── Counter_Test.json
│   │   ├── CounterTest.json
│   │   └── Counterz_Test.json
│   ├── F.sol
│   │   └── F.json
├── README.md
├── src
│   └── Counter.sol
└── test
    ├── Counter.t.sol
    └── F.sol
```

Before this fix, the `contracts.k` would look like:

[contracts.k.txt](https://github.com/runtimeverification/evm-semantics/files/12342668/contracts.k.txt)

Note that kompilation was failing because of the two `Counter` and `COuntER` with different casing, and now fails due to the `_` in 
```k
Location(40,32,40,33)
40 |        syntax Contract ::= Counter_TestContract
   .                                   ^
```
   
Now it's:

[contracts2.k.txt](https://github.com/runtimeverification/evm-semantics/files/12342679/contracts.k.txt)

 and does not fail.

### CODE REVIEW

**Comment by radumereuta** [comment]
`$` -> `ZDlr`
`_` -> `ZUnd`
`Z` -> `ZZ`
Use capital Z because it's less likely to be found in a contract.
`contract_zero` -> `KEVMcontractZUndzero`
`Contract_Zero` -> `KEVMContractZUndZZero`
`KEVM` at the end would make it more readable but runs into issues with lowercase contract names.
We can do the same for module names.

File: kevm-pyk/src/kevm_pyk/utils.py [hunk 1]
```diff
@@ -347,3 +347,53 @@ def legacy_explore(
                 id=id,
                 trace_rewrites=trace_rewrites,
             )
+
+
+def escaped_chars() -> list[str]:
```
Comment (radumereuta) [inline kevm-pyk/src/kevm_pyk/utils.py:352]: `__` make the function private.

**Comment by radumereuta** [comment]
I was thinking instead of the `KEVM` prefix, maybe `S2K` would be more appropriate because the translation is done in the `solidity-to-k` tool.

**Comment by iFrostizz** [comment]
Yes, thank you @radumereuta this makes more sense.

File: kevm-pyk/src/kevm_pyk/utils.py (hunk 1 above)
Comment (iFrostizz) [reply kevm-pyk/src/kevm_pyk/utils.py]: I've done it a `@classmethod`, is it good enough ?

**Comment by iFrostizz** [comment]
Waiting for https://github.com/runtimeverification/evm-semantics/pull/1883

File: kevm-pyk/src/kevm_pyk/utils.py (hunk 1 above)
Comment (radumereuta) [reply kevm-pyk/src/kevm_pyk/utils.py]: I'm not that familiar with Python. This is just a suggestion and is not important. Although documentation would be really nice.

Comment (iFrostizz) [reply kevm-pyk/src/kevm_pyk/utils.py]: Will do, thanks!

File: tests/foundry/lemmas.k [hunk 2]
```diff
@@ -60,8 +60,8 @@ module SUM-TO-N-INVARIANT
    andBool #rangeUInt(256, S)
    andBool GAS_AMT >=Int N *Int 178
    andBool CONDITION ==K bool2Word ( N:Int ==Int 0 )
-   andBool PROGRAM ==K #binRuntime(LoopsTest)
-   andBool JUMPDESTS ==K #computeValidJumpDests(#binRuntime(LoopsTest))
+   andBool PROGRAM ==K #binRuntime(KEVMLoopsTest)
+   andBool JUMPDESTS ==K #computeValidJumpDests(#binRuntime(KEVMLoopsTest))
```
Comment (radumereuta) [inline tests/foundry/lemmas.k:64]: KEVM here?

File: kevm-pyk/src/tests/unit/test_solc_to_k.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+from kontrol.solc_to_k import Contract
+
+TEST_DATA: list[tuple[str, str, str, str]] = [
+    ('has_underscore', 'KEVM', 'My_contract', 'KEVMMyz5fcontract'),
+    ('no_change', '', 'mycontract', 'mycontract'),
+    ('starts_underscore', 'KEVM', '_method', 'KEVMz5fmethod'),
+    ('with_escape', '', 'z_', 'z7az5f'),
+    ('with_escape', 'KEVM', 'zKEVM_', 'KEVMz7aKEVMz5f'),
+]
```
[the remark this answers is not in the document]
Comment (iFrostizz) [reply kevm-pyk/src/tests/unit/test_solc_to_k.py:13]: Not necessarily, the prefix is defined as the second argument. It's just testing for some edge cases, sometimes without a prefix

[the remark this answers is not in the document]
Comment (iFrostizz) [reply kevm-pyk/src/tests/unit/test_solc_to_k.py:13]: But for clarity I'll update to `S2K` , thank you!

File: tests/foundry/lemmas.k (hunk 2 above)
Comment (iFrostizz) [reply tests/foundry/lemmas.k:64]: Good one, thanks!

File: kevm-pyk/src/kontrol/solc_to_k.py [hunk 4]
```diff
@@ -342,17 +352,69 @@ def test_to_claim_name(t: str) -> str:
     def name_upper(self) -> str:
         return self.name[0:1].upper() + self.name[1:]
 
+    @staticmethod
+    def escaped_chars() -> list[str]:
+        return [Contract.prefix_code, '_']
+
+    @staticmethod
+    def escape_char(char: str) -> str:
+        return f'z{hex(ord(char)).removeprefix("0x")}'
```
Comment (radumereuta) [inline kevm-pyk/src/kontrol/solc_to_k.py:361]: We also agreed to make the encoding more readable by using this encoding:
```
$ -> ZDlr
_ -> ZUnd
Z -> ZZ
```

**Review by radumereuta — COMMENTED** [review]
With those minor comments, this looks fine.
But it should be reviewed by someone more familiar with the evm-semantics as well.

File: kevm-pyk/src/kontrol/solc_to_k.py [hunk 5]
```diff
@@ -342,17 +352,69 @@ def test_to_claim_name(t: str) -> str:
     def name_upper(self) -> str:
         return self.name[0:1].upper() + self.name[1:]
 
+    @staticmethod
+    def escaped_chars() -> list[str]:
+        return [Contract.prefix_code, '_']
```
[the remark this answers is not in the document]
Comment (iFrostizz) [reply kevm-pyk/src/kontrol/solc_to_k.py:357]: Thank you very much! I'll update and write tests

**Comment by iFrostizz** [comment]
Currently updating the expected output.

File: kevm-pyk/src/tests/integration/test-data/show/AssumeTest.test_assume_false(uint256,uint256).expected [hunk 6]
```diff
@@ -2263,12 +2263,7 @@ Failing nodes:
     #And { true #Equals ( notBool VV0_a_114b9705:Int ==Int VV1_b_114b9705:Int ) ==Bool true } ) ) ) ) ) ) ) ) ) ) ) #Implies { true #Equals foundry_success ( ... statusCode: EVMC_SUCCESS , failed: #lookup ( ( 46308022326495007027972728677917914892729792999299745830475596687180801507328 |-> 1 ) , 46308022326495007027972728677917914892729792999299745830475596687180801507328 ) , revertExpected: false , opcodeExpected: false , recordEventExpected: false , eventExpected: false ) } )
   Path condition:
     #Top
-  Model:
-    CALLER_ID = 0
-    NUMBER_CELL = 0
-    ORIGIN_ID = 0
-    VV0_a_114b9705 = 1
-    VV1_b_114b9705 = 0
+  Failed to generate a model.
```
Comment (anvacaru) [inline kevm-pyk/src/tests/integration/test-data/show/AssumeTest.test_assume_false(uint256,uint256).expected:2266]: Is this related to these changes?

File: kevm-pyk/src/kontrol/solc_to_k.py [hunk 7]
```diff
@@ -228,6 +237,7 @@ def application(self) -> KInner:
     raw_sourcemap: str | None
     methods: tuple[Method, ...]
     fields: FrozenDict
+    prefix_code: str = 'Z'
```
Comment (anvacaru) [inline kevm-pyk/src/kontrol/solc_to_k.py:240]: If this is supposed to be a constant, I guess it should be marked accordingly. `PREFIX_CODE: Final = 'Z'`.

File: kevm-pyk/src/kontrol/solc_to_k.py [hunk 8]
```diff
@@ -342,17 +352,86 @@ def test_to_claim_name(t: str) -> str:
...
+        """
+        escape all the chars that would cause issues once kompiling and add a prefix to mark it as "escaped"
+        """
+        escaped = prefix
+        for char in iter(name):
+            if char in Contract.escaped_chars():
+                escaped += Contract.escape_char(char)
+            else:
+                escaped += char
+        return escaped
```
Comment (anvacaru) [inline kevm-pyk/src/kontrol/solc_to_k.py:394]: Maybe something like this to have less string concatenations?
```suggestion
        escaped = [Contract.escape_char(char) if char in Contract.escaped_chars() else char for char in name]
        return prefix + ''.join(escaped)
```

File: kevm-pyk/src/kontrol/solc_to_k.py [hunk 9]
```diff
@@ -342,17 +352,86 @@ def test_to_claim_name(t: str) -> str:
...
+
+    @staticmethod
+    def unescaped(name: str, prefix: str = '') -> str:
+        if not name.startswith(prefix):
+            raise RuntimeError(f'name {name} should start with {prefix}')
+        unescaped = name.removeprefix(prefix)
+        res = ''
+        unes_iter = iter(unescaped[:-1])
+        skipped = 0
+        for i, char in enumerate(unes_iter):
```
Comment (anvacaru) [inline kevm-pyk/src/kontrol/solc_to_k.py:404]: Rather than using an iterator with explicit counting, I would iterate the string indices.

File: kevm-pyk/src/kontrol/solc_to_k.py [hunk 10]
```diff
@@ -342,17 +352,86 @@ def test_to_claim_name(t: str) -> str:
...
+            else:
+                escaped += char
+        return escaped
+
+    @staticmethod
+    def unescaped(name: str, prefix: str = '') -> str:
+        if not name.startswith(prefix):
+            raise RuntimeError(f'name {name} should start with {prefix}')
+        unescaped = name.removeprefix(prefix)
+        res = ''
```
Comment (anvacaru) [inline kevm-pyk/src/kontrol/solc_to_k.py:401]: Instead of a resulting string, perhaps you should use a list to build the result and then `join` it at the end.

File: kevm-pyk/src/tests/integration/test-data/show/AssumeTest.test_assume_false(uint256,uint256).expected (hunk 6 above)
Comment (iFrostizz) [reply kevm-pyk/src/tests/integration/test-data/show/AssumeTest.test_assume_false(uint256,uint256).expected:2266]: Most likely the local machine flakyness, thanks I missed that.

File: kevm-pyk/src/kontrol/solc_to_k.py (hunk 7 above)
Comment (iFrostizz) [reply kevm-pyk/src/kontrol/solc_to_k.py:240]: Right, thanks!

File: kevm-pyk/src/kontrol/solc_to_k.py (hunk 8 above)
Comment (iFrostizz) [reply kevm-pyk/src/kontrol/solc_to_k.py:394]: Looks good! Thanks!

File: kevm-pyk/src/kontrol/solc_to_k.py (hunk 9 above)
Comment (iFrostizz) [reply kevm-pyk/src/kontrol/solc_to_k.py:404]: I'm not sure to understand what you mean by that @anvacaru

File: kevm-pyk/src/kontrol/solc_to_k.py (hunk 10 above)
Comment (iFrostizz) [reply kevm-pyk/src/kontrol/solc_to_k.py:401]: It looks more expensive, because instead of concatenating the string on the fly, we would have to push to the array for every iteration, and then loop over every element to build this final concatenated string.

Comment (anvacaru) [reply kevm-pyk/src/kontrol/solc_to_k.py:396]: Well, strings in Python are immutable. This means that every time you concatenate two strings with `+`, a new string object is created, and the content of the two strings is copied over to the new string. This can be inefficient in a loop as you end up creating multiple intermediate objects. 

But I guess it depends on the size of the `unescaped` string. The difference may not be noticeable.

File: kevm-pyk/src/kontrol/solc_to_k.py (hunk 9 above)
Comment (anvacaru) [reply kevm-pyk/src/kontrol/solc_to_k.py:399]: Using something like  ` while i < len(unescaped):`

Comment (iFrostizz) [reply kevm-pyk/src/kontrol/solc_to_k.py:399]: Right, looks much better, thank you!

File: kevm-pyk/src/kontrol/solc_to_k.py (hunk 10 above)
Comment (iFrostizz) [reply kevm-pyk/src/kontrol/solc_to_k.py:396]: True, let's apply your suggestion. Thank you

**Review by anvacaru — APPROVED** [review]
